Added component to delay trigger events

jankrassnigg opened the pull request.

jankrassnigg committed:
Added component to delay trigger events

jankrassnigg committed:
Fixed logic

C-Core commented on Code/Engine/Core/Messages/TriggerMessage.h:
```diff
@@ -29,6 +29,8 @@ struct EZ_CORE_DLL ezMsgComponentInternalTrigger : public ezMessage
   /// Identifies what the message should trigger.
   ezHashedString m_sMessage;
 
+  ezInt32 m_iPayload = 0;
```
maybe use ezVariant for the payload?

C-Core commented on Code/Engine/Core/World/EventMessageHandlerComponent.h:
```diff
@@ -6,14 +6,47 @@
 struct ezEventMessage;
 
+/// \brief Base class for ezEventMessageHandlerComponent. This can be used if global message handling is not needed and some options shall not show up in the UI.
+class EZ_CORE_DLL ezEventMessageHandlerBaseComponent : public ezComponent
```
In our usual naming scheme we put Base at the end of the name, so this should be ezEventMessageHandlerComponent**Base**

jankrassnigg replied:
After the other refactor, we probably don't need this change anymore, anyway.

C-Core commented on Code/Engine/GameEngine/Gameplay/TriggerDelayModifierComponent.h:
```diff
@@ -0,0 +1,58 @@
+protected:
+  virtual void Initialize() override;
+
+  void OnMsgTriggerTriggered(ezMsgTriggerTriggered& msg) const;
```
This message handler doesn't need to be const

C-Core commented on Code/Engine/GameEngine/Gameplay/TriggerDelayModifierComponent.h:
```diff
@@ -0,0 +1,58 @@
+  void OnMsgTriggerTriggered(ezMsgTriggerTriggered& msg) const;
+  void OnMsgComponentInternalTrigger(ezMsgComponentInternalTrigger& msg);
+
+  mutable ezAtomicBool m_bIsActivated = false;
```
no need for atomics since queued messages are always processed single threaded

jankrassnigg committed:
fixes

C-Core committed:
Refactored event message handling

All messages can be send as event messages now and all component types can be event message handler. No need to derive from a specific type anymore.

Commit:
Applied code formatting rules

C-Core commented on Code/Engine/GameEngine/Gameplay/Implementation/TriggerDelayModifierComponent.cpp:
```diff
@@ -0,0 +1,158 @@
+        ++m_iValidActivationToken;
+
+        // store the original trigger message for later
+        m_sMessage = msg.m_sMessage;
```
msg.m_hTriggeringObject is lost

jankrassnigg replied:
Yes, because there is no 1:1 relationship anymore who entered/left a trigger.

C-Core replied:
ah yes makes sense

jankrassnigg commented on Code/Engine/Core/World/Implementation/GameObject.cpp:
```diff
@@ -895,51 +895,48 @@ void ezGameObject::PostMessageRecursive(const ezMessage& msg, ezTime delay, ezOb
-  GetWorld()->FindEventMsgHandlers(msg, this, eventMsgHandlers);
-
-  if (eventMsgHandlers.IsEmpty() == false && pSenderComponent != nullptr)
+  if (auto pEventMsg = ezDynamicCast<ezEventMessage*>(&msg))
```
I thought we'd just have overloads for ezMessage and ezEventMessage, then you wouldn't need to do dynamic casts.

C-Core replied:
I didn't want to add another variant of this method because we already have so many. The dynamic cast is super cheap in comparison to the whole search for a message handler, so we should be fine in this case.
